### PR TITLE
Bootloader infrastructure and main logic

### PR DESCRIPTION
**Pull request by kaczmarczyck**
This is the first PR containing the bootloader code. Compared to the [full code](https://github.com/kaczmarczyck/OpenSK/tree/bootloader), this PR misses the SHA256 computation using the Cryptocell.

It includes the full setup and infrastructure, including:
- desktop checks
- workflow checks
- docs updates
- linker script
- deploy code
- main logic

The bootloader directory is self-sufficient, it's basically independent from OpenSK. Therefore, I decided to give it its own `rust-toolchain` file. They currently run with the same toolchain for convenience, but an OpenSK toolchain upgrade should not necessarily burden is with also upgrading the bootloader toolchain (even though we can, of course).

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

### CODE REVIEW

File: bootloader/.cargo/config
Comment (jmichelp) [inline bootloader/.cargo/config]: You probably want to reuse the flag (`--remap-path-prefix`) we have in Tock to ensure reproducibility in the build. Otherwise the full path is being used.
Not a breaker but good for consistency :)

File: bootloader/rust-toolchain
Comment (jmichelp) [inline bootloader/rust-toolchain:2]: if this isn't the same toolchain that we use for the firmware, be careful that it requires a "hack" in the github workflow to force its installation.

File: docs/install.md
Comment (jmichelp) [inline docs/install.md]: why not using LLVM tools which are part of installed rust toolchain?
That's what Tock does.

File: bootloader/.cargo/config
Comment (kaczmarczyck) [reply bootloader/.cargo/config]: Sounds good. I introduced the flags in deploy, consistent to the other compilations.

File: docs/install.md
Comment (kaczmarczyck) [reply docs/install.md]: For me, it looks like I always need to install something, either

1. `apt install gcc-arm-none-eabi`
2. `apt install llvm`
3. `cargo install cargo-binutils && rustup component add llvm-tools` (for calling `cargo objcopy --`)

In Tock, I find usage examples of options 1 and 2. So maybe the change in the latest commit is better?

File: bootloader/rust-toolchain
Comment (kaczmarczyck) [reply bootloader/rust-toolchain:2]: I'd leave this to my future self to decide if I want to also upgrade the bootloader toolchain or change the workflows instead. The breaking workflows will be pretty hard to miss, so I think this can be left as is?

**Comment by kaczmarczyck** [comment]
I'm waiting for the documentation PR to #408 land, and then modify this PR to add its documentation to the right places.

**Comment by kaczmarczyck** [comment]
Rebase on top of the documentation changes, and added more context.

**Comment by kaczmarczyck** [comment]
Maybe the documentation should live in `customization.md`. Or link to the bootloader README. I'm taking opinions on that!

File: bootloader/README.md
Comment (jmichelp) [inline bootloader/README.md]: Is it really needed to erase UICR here?
Tock kernel may erase it again if the registers aren't set to what it expects according to the selected board.

File: deploy.py
Comment (jmichelp) [inline deploy.py:399]: I suggest to remove the flags from the `bootloader/.cargo/config` file to ensure they are only defined once.

File: docs/boards/nrf52840dk.md
Comment (jmichelp) [inline docs/boards/nrf52840dk.md]: typo: alternate

File: docs/install.md
Comment (jmichelp) [inline docs/install.md]: typos: backward and non-upgradable

File: docs/install.md
Comment (jmichelp) [inline docs/install.md]: typo: upgradable

Comment (kaczmarczyck) [reply docs/install.md]: Done

File: docs/install.md
Comment (kaczmarczyck) [reply docs/install.md]: Done

File: docs/boards/nrf52840dk.md
Comment (kaczmarczyck) [reply docs/boards/nrf52840dk.md]: Done

File: deploy.py
Comment (kaczmarczyck) [reply deploy.py:399]: Good point. README is updated to match, also setting the RUSTFLAGS manually.

File: bootloader/README.md
Comment (kaczmarczyck) [reply bootloader/README.md]: It's not, that was just a copy-paste. Changed to `--sectorerase`.

**Review by jmichelp — APPROVED** [review]

**Pull request merged**
